snap7/util.py

```diff
@@ -563,118 +586,123 @@ def get_offset(self, byte_index: Union[str, int]) -> int:
+        elif type_ == 'DATE':
             return 'read DATE not implemented'
 
-        elif _type == 'TIME_OF_DAY':
+        elif type_ == 'TIME_OF_DAY':
             return 'read TIME_OF_DAY not implemented'
 
         raise ValueError
 
-    def set_value(self, byte_index, _type, value):
+    def set_value(self, byte_index: Union[str, int], type: str, value: Union[bool, str, int, float]) -> Union[bytearray, None]:
         bytearray_ = self.get_bytearray()
 
-        if _type == 'BOOL':
+        if type == 'BOOL' and isinstance(value, bool):
             """
             mypy conform style:
             if isinstance(byte_index, str):
                 byte_index, bool_index = byte_index.split('.')
                 return set_bool(bytearray_, self.get_offset(byte_index),
                                 int(bool_index), value)
             """
-            byte_index, bool_index = byte_index.split(".")
+            byte_index, bool_index = str(byte_index).split(".")
             return set_bool(bytearray_, self.get_offset(byte_index),
                             int(bool_index), value)
 
         byte_index = self.get_offset(byte_index)
 
-        if _type.startswith('STRING'):
-            max_size = re.search(r'\d+', _type).group(0)
-            max_size = int(max_size)
-            return set_string(bytearray_, byte_index, value, max_size)
+        if type.startswith('STRING') and isinstance(value, str):
+            max_size = re.search(r'\d+', type)
+            if max_size is None:
+                raise Snap7Exception("Max size could not be determinate. re.search() returned None")
+            max_size_grouped = max_size.group(0)
+            max_size_int = int(max_size_grouped)
+            return set_string(bytearray_, byte_index, value, max_size_int)
 
-        elif _type == 'REAL':
+        elif type == 'REAL':
             return set_real(bytearray_, byte_index, value)
 
-        elif _type == 'DWORD':
+        elif type == 'DWORD' and isinstance(value, int):
             return set_dword(bytearray_, byte_index, value)
 
-        elif _type == 'DINT':
+        elif type == 'DINT' and isinstance(value, int):
             return set_dint(bytearray_, byte_index, value)
 
-        elif _type == 'INT':
+        elif type == 'INT' and isinstance(value, int):
             return set_int(bytearray_, byte_index, value)
 
-        elif _type == 'WORD':
+        elif type == 'WORD' and isinstance(value, int):
             return set_word(bytearray_, byte_index, value)
 
-        elif _type == 'USINT':
+        elif type == 'USINT' and isinstance(value, int):
+            return set_usint(bytearray_, byte_index, value)
+
+        elif type == 'SINT' and isinstance(value, int):
+            return set_sint(bytearray_, byte_index, value)
+
+        if type == 'USINT' and isinstance(value, int):
             return set_usint(bytearray_, byte_index, value)
 
-        elif _type == 'SINT':
+        if type == 'SINT' and isinstance(value, int):
```

Comment:
SINT and USINT are already listed. you can remoive these two (with just the if statement)